v2/pkg/genruntime/convertible.go

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package genruntime
+
+import (
+	"reflect"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+)
+
+func findStorageGVK(scheme *runtime.Scheme, gk schema.GroupKind) (schema.GroupVersionKind, error) {
+	interfaceType := reflect.TypeOf((*conversion.Hub)(nil)).Elem()
+
+	for gvk, rt := range scheme.AllKnownTypes() {
+		if gvk.Group != gk.Group || gvk.Kind != gk.Kind {
+			continue
+		}
+
+		if reflect.PointerTo(rt).Implements(interfaceType) {
+			return gvk, nil
+		}
```

Comment:
Why not use the existing `NewEmptyVersionedResourceFromGVK()` and then test using `if _, ok := empty.(conversion.Hub); ok`, avoiding the cryptic reflection code?

Reply:
Two reasons:
1. `scheme.AllKnownTypes` returns a `map[schema.GroupVersionKind]reflect.Type`, so it seemed to make some sense to stay in the "reflect world" when scanning it if I don't actually need an instance.
2. Since this is called in the (somewhat) hot path of the core reconcile loop logic, and can be hit at least a couple times through each loop (because there may be multiple versions of a resource for a given group + version) I was a bit concerned about performance/extra allocations. It _probably_ wouldn't matter but avoiding constructing 8-10 objects that are immediately thrown away when instead we can do a check via reflection that avoids those allocations.

I _almost_ just computed a map of this upfront once and then we could look it up whenever we wanted but that seemed overkill so instead settled on using a less-allocations approach using this reflection method that kept things "simple" (no extra variables to storage cached data that needed to be passed around) while still avoiding excess allocations.